airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/stream_reader.py

```diff
@@ -79,17 +86,92 @@ def get_matching_files(
+                    f"available memory: {available_memory / (1024 * 1024):,.2f} MB ({available_memory / (1024 * 1024 * 1024):.2f} GB)."
+                )
+
+        return progress_handler
+
+    @override
+    def get_file(self, file: RemoteFile, local_directory: str, logger: logging.Logger) -> Dict[str, str | int]:
+        """
+        Downloads a file from SFTP server to a specified local directory.
+
+        Args:
+            file (RemoteFile): The remote file object containing URI and metadata.
+            local_directory (str): The local directory path where the file will be downloaded.
+            logger (logging.Logger): Logger for logging information and errors.
+
+        Returns:
+            dict: A dictionary containing the following:
+                - "file_url" (str): The absolute path of the downloaded file.
+                - "bytes" (int): The file size in bytes.
+                - "file_relative_path" (str): The relative path of the file for local storage. Is relative to local_directory as
+                this a mounted volume in the pod container.
+
+        Raises:
+            FileSizeLimitError: If the file size exceeds the predefined limit (1 GB).
+        """
+        file_size = self.file_size(file)
+        # I'm putting this check here so we can remove the safety wheels per connector when ready.
+        if file_size > self.FILE_SIZE_LIMIT:
+            message = "File size exceeds the 1 GB limit."
+            raise FileSizeLimitError(message=message, internal_message=message, failure_type=FailureType.config_error)
+
+        file_relative_path, local_file_path, absolute_file_path = self._get_file_transfer_paths(file, local_directory)
+
+        # Get available disk space
+        disk_usage = psutil.disk_usage("/")
+        available_disk_space = disk_usage.free
+
+        # Get available memory
+        memory_info = psutil.virtual_memory()
+        available_memory = memory_info.available
+
+        # Log file size, available disk space, and memory
+        logger.info(
+            f"Starting to download the file {file.uri} with size: {file_size / (1024 * 1024):,.2f} MB ({file_size / (1024 * 1024 * 1024):.2f} GB) "
+            f"to '{local_file_path}' "
+            f"with size: {file_size / (1024 * 1024):,.2f} MB ({file_size / (1024 * 1024 * 1024):.2f} GB), "
+            f"available disk space: {available_disk_space / (1024 * 1024):,.2f} MB ({available_disk_space / (1024 * 1024 * 1024):.2f} GB),"
+            f"available memory: {available_memory / (1024 * 1024):,.2f} MB ({available_memory / (1024 * 1024 * 1024):.2f} GB)."
+        )
+        progress_handler = self.create_progress_handler(local_file_path, logger)
+        start_download_time = time.time()
+        # Copy a remote file in remote path from the SFTP server to the local host as local path.
+        self.sftp_client.sftp_connection.get(file.uri, local_file_path, callback=progress_handler)
```

Comment:
[get](https://github.com/paramiko/paramiko/blob/23f92003898b060df0e2b8b1d889455264e63a3e/paramiko/sftp_client.py#L840-L846) method calls [get_fo](https://github.com/paramiko/paramiko/blob/23f92003898b060df0e2b8b1d889455264e63a3e/paramiko/sftp_client.py#L799-L801) that calls [_transfer_with_callback](https://github.com/paramiko/paramiko/blob/23f92003898b060df0e2b8b1d889455264e63a3e/paramiko/sftp_client.py#L675-L678) that reads by 32768 bytes chunks.

Reply:
[get](https://github.com/paramiko/paramiko/blob/23f92003898b060df0e2b8b1d889455264e63a3e/paramiko/sftp_client.py#L840-L846) method calls [get_fo](https://github.com/paramiko/paramiko/blob/23f92003898b060df0e2b8b1d889455264e63a3e/paramiko/sftp_client.py#L799-L801) that calls [_transfer_with_callback](https://github.com/paramiko/paramiko/blob/23f92003898b060df0e2b8b1d889455264e63a3e/paramiko/sftp_client.py#L675-L678) that reads by 32768 bytes chunks.

Reply:
With the above, we now resolve the @artem1205 comment concern https://github.com/airbytehq/airbyte/pull/47686#discussion_r1821558810.